shared/subspace-object-fetching/src/object_fetcher.rs

```diff
@@ -0,0 +1,402 @@
+            // Not enough data without crossing segment boundary
+            return Ok(None);
+        }
+
+        let mut data = read_records_data
+            [mapping.offset as usize + data_length_bytes_length as usize..]
+            .to_vec();
+        drop(read_records_data);
+
+        // Read more pieces until we have enough data
+        while data.len() <= data_length as usize {
+            let raw_piece_bytes = self.read_and_decode_piece(next_piece_index).await?;
+            next_piece_index += PieceIndex::ONE;
+            data.extend_from_slice(&raw_piece_bytes);
+        }
+
+        // Trim the excess
+        data.truncate(data_length as usize);
+
+        Ok(Some(data))
+    }
+
+    /// Fetch and assemble an object that can cross segment boundaries, which requires assembling
+    /// and iterating over full segments.
+    async fn fetch_object_regular(&self, mapping: GlobalObject) -> Result<Vec<u8>, Error> {
+        let segment_index = mapping.piece_index.segment_index();
+        let piece_position_in_segment = mapping.piece_index.position();
+        // Used to access the data after it is converted to raw bytes
+        let offset_in_segment = (u64::from(piece_position_in_segment) * RawRecord::SIZE as u64
+            + u64::from(mapping.offset)) as usize;
+
+        let mut data = {
+            let Segment::V0 { items } = self.read_segment(segment_index).await?;
+            // Unconditional progress is enum variant + compact encoding of number of elements
+            let mut progress = 1 + Compact::compact_len(&(items.len() as u64));
+            let segment_item = items
+                .into_iter()
+                .find(|item| {
+                    // Add number of bytes in encoded version of segment item
+                    progress += item.encoded_size();
+
+                    // Our data is within another segment item, which will have wrapping data
+                    // structure, hence strictly `>` here
+                    progress > offset_in_segment
+                })
+                .ok_or_else(|| {
+                    let error_string = format!(
+                        "Failed to find item at offset {offset_in_segment} in segment \
+                        {segment_index} for object {mapping:?}"
+                    );
+                    error!(error = %error_string);
+
+                    Error::StringError(error_string)
+                })?;
+
+            match segment_item {
+                SegmentItem::Block { bytes, .. }
+                | SegmentItem::BlockStart { bytes, .. }
+                | SegmentItem::BlockContinuation { bytes, .. } => {
+                    // Rewind back progress to the beginning of the number of bytes
+                    progress -= bytes.len();
+                    // Get a chunk of the bytes starting at the position we care about
+                    Vec::from(&bytes[offset_in_segment - progress..])
```

Comment:
Do we want to check that the encoded size is smaller than MAX_OBJECT_SIZE before possibly retrieving another segment?

Reply:
I don't think I understood your comment fully. Are you suggesting to check `MAX_OBJECT_SIZE` to prevent the unexpected object size? If so, I would say - yes, because we need to stop somewhere (in theory, we can have an endless object spanning across all the segments).